### PR TITLE
rocblas: disable Beta API on Windows for HIP < 5.7

### PR DESCRIPTION
**Pull request by apwojcik**
The official HIP SDK release for Windows is 5.5.1. The Beta API is not available in that release. The API will be available from 5.7 and above on Windows.

### CODE REVIEW

File: src/gemm_v2.cpp
Comment (JehandadKhan) [inline src/gemm_v2.cpp]:
```suggestion
#if ROCBLAS_BETA_FEATURES_API
```

**Review by JehandadKhan — CHANGES_REQUESTED** [review]

**Comment by junliume** [comment]
@apwojcik could you follow up with the change request above?

**Comment by apwojcik** [comment]
I did, and the change was made. I am waiting for @JehandadKhan to review it again.

File: src/gemm_v2.cpp
Comment (JehandadKhan) [inline src/gemm_v2.cpp]: @apwojcik  IIUC this means the `ROCBLA_BETA_FATURES_API` would always be enabled on Linux, regardless of the HIP version, which is not what we want ?

**Review by JehandadKhan — CHANGES_REQUESTED** [review]

File: src/gemm_v2.cpp
Comment (JehandadKhan) [inline src/gemm_v2.cpp]:
```suggestion
#if defined(_WIN32) || (HIP_PACKAGE_VERSION_FLAT < 5007000000ULL)
```

Comment (JehandadKhan) [reply src/gemm_v2.cpp]: @apwojcik  Would you agree? Perhaps I am missing something.

**Review by JehandadKhan — DISMISSED** [review]

File: src/gemm_v2.cpp
Comment (apwojcik) [reply src/gemm_v2.cpp]: There was no check before and on Linux Beta Feature API was always on. I made it optional on Windows and available only for HIP SDK 5.7, but not for HIP SDK 5.5 (which is officialy released on Windows and there is no newer available).

Comment (apwojcik) [reply src/gemm_v2.cpp]: I can change as you suggesting so all using older HIP SDK will compile on Linux, too.

Comment (apwojcik) [reply src/gemm_v2.cpp]: @JehandadKhan do you want me to change that for Linux, too? By default, Linux installs the latest available ROCm - currently 5.7

File: src/gemm_v2.cpp
Comment (junliume) [inline src/gemm_v2.cpp]: @apwojcik minor CI issues since we treat warnings as errors:
```
/var/jenkins/workspace/MLLibs_MIOpen_roblas_beta_api/src/gemm_v2.cpp:121:5: error: 'ROCBLAS_BETA_FEATURE_API' is not defined, evaluates to 0 [-Werror,-Wundef]

#if ROCBLAS_BETA_FEATURE_API

    ^
```

Comment (atamazov) [inline src/gemm_v2.cpp]: Typo.

**Review by atamazov — CHANGES_REQUESTED** [review]

**Comment by atamazov** [comment]
@apwojcik @JehandadKhan @junliume There are some important questions that need to be answered before this PR is merged in.

Some considerations first.

AFAICS `ROCBLAS_BETA_FEATURES_API` exposes some "beta" API entities in the roclblas interface. In rocblas 3.1.0 it includes:

- `rocblas_gemm_ex_get_solutions()`
- `rocblas_gemm_ex_get_solutions_by_type()`
- `rocblas_gemm_batched_ex_get_solutions()`
- `rocblas_gemm_batched_ex_get_solutions_by_type()`
- `rocblas_gemm_strided_batched_ex_get_solutions()`
- `rocblas_gemm_ex3()`

However the list of "beta" entities is not permanent. In some future versions or rocblas the beta API stuff may change. For example, some of the beta stuff may be moved to the normal rocblas API. Some new entities may appear under the ROCBLAS_BETA_FEATURES_API guard. Some existing beta things may be changed or removed etc. 

**_When we define ROCBLAS_BETA_FEATURES_API as 1 (prior inclusion of `rocblas.h`), the only meaning of that is that we would like to get access to all the beta features of rocblas._** Defining `ROCBLAS_BETA_FEATURES_API` prior inclusion of `rocblas.h` should be allowed with any rocblas version and can't lead to build errors.

However the list of beta features is not known and the only way to find them out is examining the rocblas version. For example, please see how we define and use `USE_ROCBLAS_GEMM_EX3` in PR #2480.

> The official HIP SDK release for Windows is 5.5.1. The Beta API is not available in that release. The API will be available from 5.7 and above on Windows.

What is the exact problem this PR is intended to resolve?
- If some "beta" feature is not available on Windows yet, then we need to examine rocblas version and disable the usage of that feature based on that. Usage of HIP version is not Ok because developers and QA may install some different rocblas on top of specific HIP SDK version.
- If rocblas versioning on Windows and Linux is different, then we need to take that into account. 

----
[Note] As we can see, this “ROCBLAS_BETA_FEATURES_API” is almost useless in practice. Its only function is to hide beta functions from the user. That is why I suggested not simply reusing similar approach in MIOpen, but [extend it with features that indicate the presence of a particular beta function and even report its version number](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2320#discussion_r1316419544).

**Comment by atamazov** [comment]
@apwojcik BTW please check if #2480 resolves the problem on Windows.

**Review by atamazov — CHANGES_REQUESTED** [review]
No go. Please see https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2405#issuecomment-1787866776. It is likely that #2480 supersedes this PR (but I an not sure because I do not know what exactly is the problem with Windows).

File: src/include/miopen/handle.hpp
Comment (atamazov) [inline src/include/miopen/handle.hpp]: [Side note] @JehandadKhan BTW is seems like handle.hpp does not use any beta API features from rocblas. Shall we simply remove ROCBLAS_BETA_FEATURES_API here?

Comment (JehandadKhan) [reply src/include/miopen/handle.hpp]: Yes, please

Comment (atamazov) [reply src/include/miopen/handle.hpp]: @JehandadKhan Oh, no. We have to define ROCBLAS_BETA_FEATURES_API in each header where rocblas.h is included. Because the order of inclusion of headers is not guaranteed, and rocblas.h in included only once, at the first appearance!

Comment (atamazov) [reply src/include/miopen/handle.hpp]: @JehandadKhan ...and, as gemm_v2.cpp includes handle.hpp, removal of ROCBLAS_BETA_FEATURES_API from handle.hpp would lead to that rocblas.h is included (from handle.hpp) without beta stuff.

Comment (atamazov) [reply src/include/miopen/handle.hpp]: @JehandadKhan 
> We have to define ROCBLAS_BETA_FEATURES_API in each header where rocblas.h is included

I've quickly inspected the code and found that rocblas.h is included only in handle.hpp and gemm_v2.cpp, so we are fine.

[end of the side thread]

**Comment by apwojcik** [comment]
> @apwojcik BTW please check if #2480 resolves the problem on Windows.

It does partially resolve. However, for HIP SDK 5.5.1 on Windows, we still need to switch the beta API off since there is no `rocblas_computetype` available. It is available starting from HIP SDK 5.7

**Comment by atamazov** [comment]
@apwojcik 
> It does partially resolve. However, for HIP SDK 5.5.1 on Windows, we still need to switch the beta API off since there is no `rocblas_computetype` available. It is available starting from HIP SDK 5.7

Can you please let me know which rocblas version is included into HIP SDK 5.5.1 and into HIP SDK 5.7? It should be 2.47 and 3.0 (I guess).

**Comment by atamazov** [comment]
@apwojcik Please read https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2405#issuecomment-1787866776 and do not hesitate to ask if something left unclear. It explains why this PR should not be merged in. Sorry.

As far as I understand, Wnidows HIP SDK 5.5.1 contains rocblas 2.47 that do not support GEMM_EX3 feature, while rocblass 2.47 in ROCm supports it. Let me update #2480.

**Comment by atamazov** [comment]
@apwojcik As #2480 is merged in, then let's make the necessary changes right here:
1. Revert everything in this branch
2. Update this branch from the fresh `develop`
3. In src/gemm_v2.cpp, replace
```C++
#define USE_ROCBLAS_GEMM_EX3 ((MIOPEN_ROCBLAS_VERSION_FLAT >= 2047000) && ROCBLAS_BETA_FEATURES_API)
```
with
```C++
#ifdef _WIN32
#define USE_ROCBLAS_GEMM_EX3 ((MIOPEN_ROCBLAS_VERSION_FLAT >= 3000000) && ROCBLAS_BETA_FEATURES_API)
#else
#define USE_ROCBLAS_GEMM_EX3 ((MIOPEN_ROCBLAS_VERSION_FLAT >= 2047000) && ROCBLAS_BETA_FEATURES_API)
#endif
```
~Profit!~ That should work for you.

**Comment by apwojcik** [comment]
@atamazov it's done

**Review by atamazov — APPROVED** [review]
LGTM!

**Comment by atamazov** [comment]
@apwojcik 
> force-pushed the roblas_beta_api branch from 76b249a to d3a4ed2

No problem in this PR. But just in case if you are not aware that force-pushing is not welcome here:

<details><summary><h3>⚠️ Please do not force-push (click to see why)</h3></summary><hr>

This is NOT recommended due to several reasons. For example, this makes previous code reviews invisible on the "Code changes" tab (due to overwritten commit hashes). Force-pushing can only be used more or less safely if the branch has NOT yet been published for review.

In order to resolve merge conflicts it is recommended to _merge `develop`_ into your branch. _Rebasing on top of `develop`_ leads to repeating merge conflicts (because **_we squash branches_** prior merging into `develop`) and therefore not recommended.

If develop has been substantially updated since branch creation, then it is recommended to update from `develop` even if there are no merge conflicts. This is to ensure that recent changes to `develop` are actually compatible with your changes. Otherwise, `develop` may become broken after the merge.
<hr></details>

**Comment by atamazov** [comment]
@apwojcik BTW does this PR resolve the issue for you? Thanks!

**Comment by apwojcik** [comment]
> @apwojcik BTW does this PR resolve the issue for you? Thanks!

Yes, it does. Thank you for your effort.

**Comment by CAHEK7** [comment]
@junliume shall we update to the lates develop? It seems the change is trivial and affects only windows build.

**Review by CAHEK7 — APPROVED** [review]

**Review by junliume — APPROVED** [review]

**Pull request merged**
